### ISSUE TITLE
Serialisation of DateTime when CultureInvariant is specified does not result in truly invariant output for clients

### ISSUE THREAD

**Issue by NeilMacMullen** [JoshClose/CsvHelper#1467]
**Describe the bug**
Serialising a DateTime such as "1 oct 2019" results in a string such as 

> 10/01/2019 00:00:00

A significant problem with this serialisation is that it is NOT treated as invariant by Excel/PowerBi and probably other consumers of the CSV since they will use the local culture by default and may therefore incorrectly interpret the month/day order. I.e. Excel may treat this date as "10th Jan 2019" on a UK machine.

It would be preferable to use a format which is less ambiguous such as 

> "1 october 2019 00:00:00" 
> or 
>  "2019-10-01 00:00:00"

Something like https://en.wikipedia.org/wiki/ISO_8601 would be ideal but it doesn't seem that Excel recognises this :-( 

As a workaround I currently avoid DateTimes in my classes and do my own serialisation to strings  (guess I could use TypeConversion as well).

Anyway, great library  - I use it all the time but this issue does trip me up quite a bit since we often transfer csvs between US and UK

**Comment by NeilMacMullen** [comment]
It appears a sufficient workaround is to use this construction...
```

  using (var writer = new StreamWriter(...))
  {
      var trulyInvariant = (CultureInfo) CultureInfo.InvariantCulture.Clone();
      trulyInvariant.DateTimeFormat.ShortDatePattern = "yyyy-MM-dd";
      using (var csv = new CsvWriter(writer, trulyInvariant))
      {
             csv.WriteRecords(records);
      }
  }
```

**Comment by JoshClose** [comment]
There is no specified format in CSV, so you can write it however you want. Invariant is the way .NET handles being culture agnostic. Excel really has nothing to do with CSV. If you want to write something that people will be reading in Excel, then you can configure the format to whatever Excel likes.

Here is configuring for ISO 8601.

You can do it as an attribute like this.

```cs
[Format("o")]
public DateTimeOffset Date { get; set; }
```

Or as part of a class map like this.

```cs
Map(m => m.Date).TypeConverterOption.Format("o");
```

**Comment by NeilMacMullen** [comment]
Maybe classing this as a bug is a bit strong but my point is that the current configuration mechanism doesn't offer a "pit of success" option for ensuring the best chance of transferability between regions.  Saying "Excel has nothing to do with CSV" seems a bit purist - the reality is that many CSVs produced by CsvHelper are likely to be consumed by client applications (Excel/Google sheets/OpenOffice/PowerBi/Tableau/R) whose first guess will be that that the CSV is in the local culture.  Using an ambiguous date format is therefore a therefore a bad thing.  (Yes, I get that there are other incompatibilities such as number formats but dates are the primary one between US/UK and it's not always apparent from looking at a few samples what source format was used.)

Of course it's your library :-)  but for those of us who want it to "just work" as much as possible it strikes me that providing a static CultureUSUKTransferable member in CsvHelper would be helpful - no need then to have to remember to decorate/map every member that might be a Date.

**Comment by JoshClose** [comment]
This is part of the reason I added the required `CutlureInfo` parameter. It makes people think how they actually want it written and read. What about people that need to have files work between the US and other countries, or the UK and other countries? Their formats will be completely different. I don't know how Excel behaves in those other languages. `InvariantCulture` is a way for a file to be written anywhere in the world, and then read from any other place in the world.

Excel has nothing to do with CSV and I'm not really interested in doing Excel specific work for this library. I do however have a contrib organization and library set up if people would like to contribute some Excel specific things. No one has wanted to do the work yet though. https://github.com/CsvHelperContrib/CsvHelperContrib

**Comment by JoshClose** [comment]
If you look at the issues list, most are Excel related. https://github.com/CsvHelperContrib/CsvHelperContrib/issues

**Comment by NeilMacMullen** [comment]
>What about people that need to have files work between the US and other countries, or the UK and other countries?

Yes - that's why I suggested calling the 'canned' CultureInfo "USUKTransferable" or some variant - my use of "trulyInvariant" in the original posting was thoughtless :-)  I'm not trying to solve the global problem here, just pointing out there is a needless and easily avoidable incompatibility between two signifcant regions that can be fixed with very little effort!    

>This is part of the reason I added the required CutlureInfo parameter

Yes - understand why you would use CultureInfo - it's just unfortunate that CultureInvariant which many .net developers will associate with culture-independence/transferability (between .Net programs) took the short-sighted approach of using US formatting for dates rather than some clearer standard.


>Excel has nothing to do with CSV and I'm not really interested in doing Excel specific work

I don't blame you for wanting to avoid Excel-idiosyncracies!   This isn't an Excel-specific problem though - any client application or indeed human reader in the US or UK will have a hard time "guessing" whether dates were generated as dd-mm-yyyy or mm-dd-yyyy in a csv generated using CultureInvariant.

 > I do however have a contrib organization and library set up if people would like to contribute

I notice there isn't any code in that - are you happy for people just to create a new solution and projects?  I'd be happy to contribute some code but getting the entire build chain set up to generate a new package is quite a high hurdle for a few lines...

**Comment by JoshClose** [comment]
I could create a solution. There will probably be several projects in it. `CsvHelperContrib.Excel` would probably be one to handle the Excel specific things. I could do that some time today if you want to contribute code.

**Comment by NeilMacMullen** [comment]
There's no rush for me - I have already created a local library to solve my particular problem and realistically am busy on other things until next week but will contribute if you get something in place :-)
